commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoDocumentFilter.scala

```diff
@@ -0,0 +1,137 @@
+      if (doc.containsKey(op.rawOperator)) {
+        throw new IllegalArgumentException(s"duplicate query operator ${op.rawOperator}")
+      }
+      doc.put(op.rawOperator, op.toBson)
+    }
+    doc
+  }
+}
+
+sealed trait MongoDocumentFilter[E] extends MongoFilter[E] {
+
+  import MongoDocumentFilter._
+
+  def and(other: MongoDocumentFilter[E]): MongoDocumentFilter[E] = (this, other) match {
+    case (Empty(), other) => other
+    case (thiz, Empty()) => thiz
+    case (And(thisFilters), And(otherFilters)) => And(thisFilters ++ otherFilters)
+    case (And(thisFilters), other) => And(thisFilters :+ other)
+    case (thiz, And(otherFilters)) => And(thiz +: otherFilters)
+    case _ => And(Vector(this, other))
+  }
+
+  def &&(other: MongoDocumentFilter[E]): MongoDocumentFilter[E] = and(other)
+
+  def or(other: MongoDocumentFilter[E]): MongoDocumentFilter[E] = (this, other) match {
+    case (Empty(), _) | (_, Empty()) => empty
+    case (Or(thisFilters), Or(otherFilters)) => Or(thisFilters ++ otherFilters)
+    case (Or(thisFilters), other) => Or(thisFilters :+ other)
+    case (thiz, Or(otherFilters)) => Or(thiz +: otherFilters)
+    case _ => Or(Vector(this, other))
+  }
+
+  def ||(other: MongoDocumentFilter[E]): MongoDocumentFilter[E] = or(other)
+
+  final def toFilterBson(prefixPath: Opt[String]): BsonDocument = {
+    val docs = new BsonArray
+    addToFilters(prefixPath, docs)
+    docs.size match {
+      case 0 => Bson.document()
+      case 1 => docs.get(0).asDocument
+      case _ => Bson.document(Bson.And, docs)
+    }
+  }
+
+  private def addToFilters(prefixPath: Opt[String], filterDocs: BsonArray): Unit = {
+    def fullPath(suffix: String): String =
+      prefixPath.fold(suffix)(_ + "." + suffix)
+
+    def findFilterDoc(path: String): BsonDocument =
+      filterDocs.iterator.asScala.map(_.asDocument)
+        .findOpt(doc => !doc.containsKey(path) && doc.keySet.asScala.forall(k => !k.startsWith("$")))
+        .getOrElse(Bson.document().setup(filterDocs.add))
+
+    this match {
+      case Empty() =>
+      case And(filters) => filters.foreach(_.addToFilters(prefixPath, filterDocs))
+      case Or(filters) => filterDocs.add(Bson.document(Bson.Or, Bson.array(filters.iterator.map(_.toFilterBson(prefixPath)))))
+      case Nor(filters) => filterDocs.add(Bson.document(Bson.Nor, Bson.array(filters.iterator.map(_.toFilterBson(prefixPath)))))
+
+      case PropertyValueFilter(prop, filter) => filter match {
+        case docFilter: MongoDocumentFilter[_] =>
+          docFilter.addToFilters(fullPath(prop.propertyPathString).opt, filterDocs)
+
+        case opFilter: MongoOperatorsFilter[_] =>
+          val path = fullPath(prop.propertyPathString)
+          findFilterDoc(path).put(path, opFilter.toOperatorsBson)
+      }
+    }
+  }
+}
+object MongoDocumentFilter {
+  private[this] val reusableEmpty = Empty()
+
+  def empty[E]: MongoDocumentFilter[E] = reusableEmpty.asInstanceOf[MongoDocumentFilter[E]]
```

Comment:
A filter that's never satisfied is often useful as well (e.g. fold + or). I think `Filters.exists("_id", false)` is always correctly recognized as useless by Mongo when executing such queries, doesn't break index use etc.

Reply:
`Filters.exists("_id", false)` will not work for embedded documents.

Due to various random limitations of MongoDB queries I decided to write API only for things which are straightforward to express. Later we may think about emulating some lacking elements like "always false" filter.

Reply:
That's fine by me, I'd just love to have a way to construct a generic method returning such `MongoDocumentFilter[T]` which is probably impossible on the library client side (without taking a ref / companion arg)